fetch.bs

```diff
@@ -2353,6 +2457,310 @@ X-Content-Type-Options           = "nosniff" ; case-insensitive</pre>
+ <a abstract-op lt="Compute a Token Binding ID">computing a token binding ID</a>
+from <var>providedTokenBindingKeyPair</var> and
+<var>providedTokenBindingKeyParameters</var>.
+
+ <li><p>Let <var>providedTokenBinding</var> be the result of computing a <a for=/>token binding</a>
+ using <var>providedTokenBindingId</var> and <a for=/>token-binding type</a>
+ <code>provided_token_binding</code>,
+ containing a signature (using <var>providedTokenBindingKeyPair</var>) over the
+ <var>providedTokenBindingId</var> as well as <var>tlsConnection</var>'s
+ <a for=connection>token-binding Exported Keying Material</a>, as described in
+ <a href="https://tools.ietf.org/html/draft-ietf-tokbind-protocol#section-3.3">section 3.3</a>
+ of [[!TOKBIND-PROTOCOL]].
+
+ <li><p>Set <var>tokenBindingMessage</var> to a list containing <var>providedTokenBinding</var>
+ as the only <a for=/>token binding</a>.
+
+ <li><p>If <var>httpRequest</var>'s <a for=request>referred-token-binding origin</a> is not null,
+ then run these substeps:
+  <ol>
+   <li><p>Let <var>referredTlsConnection</var> be the result of
+   <a lt="obtain a connection" for=connection>obtaining a connection</a> for
+   <var>httpRequest</var>'s <a for=request>referred-token-binding origin</a>
+   with credentials <code>true</code> from the <a>connection pool</a>.
+
+   <li><p>If <a for=connection>token-binding protocol version</a> in
+   <var>referredTlsConnection</var> is null,
+   set <var>referredTokenBinding</var> to null.
+
+   <li><p>If <a for=connection>token-binding protocol version</a> in
+   <var>referredTlsConnection</var> is not null then:
+    <ol>
+     <li><p>Let <var>referredTokenBindingKeyParameters</var> be
+     <var>referredTlsConnection</var>'s <a for=connection>token-binding key parameters</a>.
+
+     <li><p>Let <var>referredTokenBindingKeyPair</var> be the result of
+     <a abstract-op lt="Get the token-binding key">getting the token-binding key</a>
+     for <var>httpRequest</var>'s <a for=request>referred-token-binding origin</a>
+     and <var>referredTokenBindingKeyParameters</var>.
+
+     <li><p>Let <var>referredTokenBindingId</var> be the result of
+     <a abstract-op lt="Compute a Token Binding ID">computing a token binding ID</a>
+     from <var>referredTokenBindingKeyPair</var> and
+     <var>referredTokenBindingKeyParameters</var>.
+
+     <li><p>Let <var>referredTokenBinding</var> be the result of computing a
+     <a for=/>token binding</a> using <var>referredTokenBindingId</var> and
+     <a for=/>token-binding type</a>
+     <code>referred_token_binding</code>, containing a signature
+     (using <var>referredTokenBindingKeyPair</var>) over the <var>referredTokenBindingId</var>
+     as well as <var>tlsConnection</var>'s
+     <a for=connection>token-binding Exported Keying Material</a>, as described in
+     <a href="https://tools.ietf.org/html/draft-ietf-tokbind-protocol#section-3.3">section 3.3</a>
+     of [[!TOKBIND-PROTOCOL]].
+    </ol>
```

Comment:
Ditto about this being poorly defined. Instead of effectively repeating the substeps from above, would it make sense to define a routine that takes a key, a type (provided or referred), and an EKM value, and returns a TokenBinding struct that can be used here and above? Perhaps replace the "Computing a token binding ID" routine with one that computes a TokenBinding struct (or maybe its serialization)?

Reply:
I added a "Computing a token binding" section.